### ISSUE TITLE
Change default form action errors when progressively enhanced

### ISSUE THREAD

**Issue by arackaf** [sveltejs/kit#7956]
### Describe the problem

I hope I can convince the maintainers that the default behavior for errors in form actions should be changed. 

Currently, if you have a progressively enhanced form (without a custom callback), and it errors, your entire page unloads, and displays the nearest error boundary (if present) or shows a default error page if not.

Errors happen, but there's no way this should be the default behavior. A user should not see their current page unloaded, and replaced with an error page because a form post happened to fail.

### Describe the proposed solution

For progressively enhanced forms, if the action fails, leave the form in place, and send the error info into $page.form, and rely on the user to display whatever error ui they'd like.

Anticipating one possible response, this would cause use:enhance to break with default, non-enhanced behavior. If that's a big problem then I would submit (pun!) that a non-enhanced form should probably not redirect to an error page either, if there's an error, but rather re-render the same form, with info about the error, so the user can potentially try again. I led with enhanced forms only since that's what I care about most, and figured that would make this a smaller request; but changing the default behavior of form action errors might be the simpler (and better) solution here. 

### Alternatives considered

The current workaround is to return a custom callback, and simply not call update(). 

That works, but it basically makes use:enahnce (by itself) essentially useless, unless you can guarantee that the action will never error out (or you disagree with my premise, and actually want the entire page to unload, and show an error page). 

### Importance

would make my life easier

**Comment by arackaf** [comment]
I don't expect maintainers to design api's around Twitter polls, but I do feel strongly about this (and really want SvelteKit to succeed) so I opened a poll here to try to gauge opinions beyond just my own 

https://twitter.com/AdamRackis/status/1599899980528951297

**Comment by PatrickG** [comment]
I'm against this change as well.

As you wrote, you can easily implement this on your own by passing a callback to the `enhance` action or create your own `enhance` action.
If you want this behavior for "un-enhanced"-forms as well, you can wrap your form action in a try-catch block and `return invalid()` in the catch block.

**Comment by arackaf** [comment]
@matt-l-w it seems crazy to me that an application framework like SvelteKit should be beholden to how web forms happen to work by default, given a web platform that was designed decades ago, to serve static documents.

To say nothing of the fact that implementing user-hostile defaults will not fix those standards. Folks who maintain these standards historically believe in “don’t break the web” (a good thing) so it’s not like frameworks implementing different defaults will somehow get these standards changed. 

And a number of people replied to tell me that popping the user to an error page goes against a11y.

**Comment by Rich-Harris** [comment]
Error pages aren't happy-path UI. Outside 404s, if you see one it's almost always because you, the application developer, fucked up somewhere. In this case, you're not handling unexpected errors inside your action. There are basically three possible reasons an action could fail:

1. user data was invalid, in which case `return invalid(4xx, {...})`
2. something went wrong on the server (an external API failed, or whatever) in which case `return invalid(5xx, {...})`
3. something went wrong on the server but you failed to anticipate it <-- YOU ARE HERE

(Side-note, but I wonder if `invalid` should be renamed `fail` or something, since 'invalid' feels more 4xx-y than 5xx-y.)

What you're saying, in effect, is that you can't possibly be expected to anticipate all the ways in which your app should go wrong, and so the framework should step in to save you from yourself — even though it has no special knowledge of the error — _but only in the case of progressively-enhanced forms_ (not `load`, or references to missing data in your markup, or a manual `fetch` to a non-existent URL, etc etc etc). Handling errors is _your_ job. The framework's `+error.svelte` boundaries are more of a backstop than anything.

If you don't handle the error and expect the framework to do (handwaves) _something_ when it occurs, then _maybe_ you've made the progressive enhancement case slightly nicer, but your non-progressively-enhanced forms (whether because you didn't add `use:enhance` or because JS failed for some reason) remain broken.

And that's only if you add code to handle the error case on the page. Let's talk about that. SvelteKit can't help you in this case...

```svelte
<form method="POST" use:enhance action="?/newTodo">
  <input name="description" placeholder="What do you need to do?">
</form>
```

...because there's nowhere to show an error. We could add some UI there...

```diff
<form method="POST" use:enhance action="?/newTodo">
+  {#if $page.error}
+    <p class="error">{$page.error.message}</p>
+  {/if}
  <input name="description" placeholder="What do you need to do?">
</form>
```

...but what happens when you have multiple forms on the page? Where does the `{#if $page.error}` UI go here?

```svelte
{#each data.todos as todo}
  <form method="POST" action="?/updateTodo" class:done={todo.done}>
    <input hidden name="id" value={todo.id}>
    <input name="description" value={todo.description}>
    <button formaction="?/toggleTodo>{todo.done ? 'mark todo' : 'mark done'}</button>
  </form>
{/each}
```

If you're handling the error yourself, you can include the `id` of the todo when you return `invalid(...)` to create the appropriate UI. The framework can't possibly be expected to have that level of understanding, because _that's not its job_ — it's yours.

In summary, the current behaviour is deliberate and isn't going to change. Please don't use context-free Twitter polls with leading questions to try and influence maintainers — we're smarter than that.

**Comment by arackaf** [comment]
I did clarify that I'd be happy to see non PE forms change this default behavior, too.

There's a variation of option 2 that's not accounted for, above. If your network is spotty, and for whatever reason does not connect to your form's endpoint, your existing UI, iiuc, will unload and render an error page, rather than update some sort of $page.form.error, (or perhaps $page.form[id].error if there's more than one form)

More broadly, it sounds like your UI will break badly unless app developers catch every possible exception, and wrap with invalid(). 

Why isn't that the default, then? Shouldn't app frameworks try to put devs into a "pit of success"? Defaults matter.
